File: _typescript/modules/libs/workspace/adapters/json/system/src/lib/session/strategy.ts
import { Session, Strategy as SystemStrategy } from '@allors/workspace/adapters/system';
import { Class, Origin } from '@allors/workspace/meta/system';
import { DatabaseOriginState } from './originstate/DatabaseOriginState';
import { DatabaseRecord } from '../database/DatabaseRecord';

export class Strategy extends SystemStrategy {
  constructor(public session: Session, public cls: Class, public id: number) {
    super(session, cls, id);

    if (this.cls.origin === Origin.Database) {
      this.DatabaseOriginState = new DatabaseOriginState(this, session.workspace.database.getRecord(this.id));
    }
  }

  static fromDatabaseRecord(session: Session, databaseRecord: DatabaseRecord) {
    return new Strategy(session, databaseRecord.cls, databaseRecord.id);
  }
}

File: dotnet/System/Workspace/Adapters/Allors.Workspace.Adapters.Remote/Session/Strategy.cs
// <copyright file="Object.cs" company="Allors bv">
// Copyright (c) Allors bv. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using Meta;

    public sealed class Strategy : Adapters.Strategy
    {
        internal Strategy(Adapters.Session session, IClass @class, long id) : base(session, @class, id) => this.DatabaseOriginState = new DatabaseOriginState(this, (DatabaseRecord)((DatabaseConnection)session.Workspace.DatabaseConnection).GetRecord(this.Id));

        internal Strategy(Session session, DatabaseRecord databaseRecord) : base(session, databaseRecord) => this.DatabaseOriginState = new DatabaseOriginState(this, databaseRecord);

        public new Session Session => (Session)base.Session;
    }
}
